controller/uninstall_controller.go

```diff
@@ -0,0 +1,437 @@
+	engineImageInformer lhinformers.EngineImageInformer,
+	nodeInformer lhinformers.NodeInformer,
+	daemonSetInformer v1beta2.DaemonSetInformer,
+) *UninstallController {
+	c := &UninstallController{
+		namespace:                 namespace,
+		force:                     force,
+		firstRun:                  true,
+		ds:                        ds,
+		stopCh:                    stopCh,
+		volumeInformerSynced:      volumeInformer.Informer().HasSynced,
+		engineInformerSynced:      engineInformer.Informer().HasSynced,
+		replicaInformerSynced:     replicaInformer.Informer().HasSynced,
+		engineImageInformerSynced: engineImageInformer.Informer().HasSynced,
+		nodeInformerSynced:        nodeInformer.Informer().HasSynced,
+		daemonSetInformerSynced:   daemonSetInformer.Informer().HasSynced,
+		queue:                     workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "longhorn-uninstall"),
+	}
+
+	volumeInformer.Informer().AddEventHandler(c.controlleeHandler())
+	engineInformer.Informer().AddEventHandler(c.controlleeHandler())
+	replicaInformer.Informer().AddEventHandler(c.controlleeHandler())
+	engineImageInformer.Informer().AddEventHandler(c.controlleeHandler())
+	nodeInformer.Informer().AddEventHandler(c.controlleeHandler())
+	daemonSetInformer.Informer().AddEventHandler(c.controlleeHandler())
+	return c
+}
+
+func (c *UninstallController) controlleeHandler() cache.ResourceEventHandler {
+	return cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(obj interface{}) { c.enqueueControlleeChange() },
+		UpdateFunc: func(old, cur interface{}) { c.enqueueControlleeChange() },
+		DeleteFunc: func(obj interface{}) { c.enqueueControlleeChange() },
+	}
+}
+
+func (c *UninstallController) enqueueControlleeChange() {
+	c.queue.AddRateLimited("uninstall")
+}
+
+func (c *UninstallController) Run() {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	logrus.Infof("Uninstalling...")
+
+	if !controller.WaitForCacheSync("longhorn uninstall", c.stopCh,
+		c.volumeInformerSynced, c.engineImageInformerSynced,
+		c.nodeInformerSynced, c.daemonSetInformerSynced) {
+
+		logrus.Fatalf("Failed to sync informers")
+	}
+
+	go wait.Until(c.worker, time.Second, c.stopCh)
+	<-c.stopCh
+	logrus.Infof("Uninstallation complete")
+}
+
+func (c *UninstallController) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *UninstallController) processNextWorkItem() bool {
+	key, quit := c.queue.Get()
+
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+
+	err := c.uninstall()
+	c.handleErr(err, key)
+
+	return true
+}
+
+func (c *UninstallController) handleErr(err error, key interface{}) {
+	if err == nil {
+		c.queue.Forget(key)
+		return
+	}
+
+	logrus.Warnf("Error: %v", err)
+	c.queue.AddRateLimited(key)
+}
+
+func (c *UninstallController) uninstall() error {
+	if c.firstRun {
+		if err := c.checkReadyShutdown(); err != nil {
+			return err
+		}
+		c.firstRun = false
+	}
+
+	if done, err := c.deleteCRDs(); err != nil {
```

Comment:
Just return err when it's not done. It's confusing to return both bool and error.

Same below.

And in the function, you can just continue to the next item if one item has been cleaned up without error. That makes logic simpler.

Reply:
Ok... I made this change, but it muddies the concept of an error. Usually, if there's some error (ex: resource changed conflicts), we want to requeue. If we _always_ return an error (until controller reaches completion), we always requeue, causing infinite queue loop.

Therefore, we must discard the error and never requeue. I've made that change in latest review.

Reply:
I think we need to introduce a `status` field in controller, which is set when the controller loop reaches a wait state (worker should terminate, awaiting change event). I'm going to propose the change with next code review - it should address 3 or 4 of your most recent comments.